feat: Instructor course information v2 apis

jawad-khan opened the pull request:
## Description

This PR implements the first two endpoints mentioned in [issue #37465](https://github.com/openedx/edx-platform/issues/37465).

These endpoints are designed for integration with the [Instructor MFE](https://github.com/openedx/frontend-app-instruct).
Since the plan is to deprecate the existing instructor APIs and replace them with these new versions, this implementation introduces a new v2 API version under the instructor app.

The decision to add these under version v2 ensures:

- Backward compatibility for existing clients.
- A smooth migration path for MFEs consuming the new API.
- Easier future maintenance and the ability to deprecate older endpoints gradually.


### Endpoints Added

Api url:
`api/instructor/v2/courses/course-v1:edX+DemoX+Demo_Course`
Response value:
`

{
    "course_id": "course-v1:edX+DemoX+Demo_Course",
    "display_name": "Demonstration Course",
    "org": "edX",
    "course_number": "DemoX",
    "enrollment_start": "2013-02-05T00:00:00Z",
    "enrollment_end": null,
    "start": "2013-02-05T05:00:00Z",
    "end": null,
    "pacing": "instructor",
    "has_started": true,
    "has_ended": false,
    "total_enrollment": 4,
    "enrollment_counts": {
        "audit": 3,
        "verified": 1,
        "total": 4
    },
    "num_sections": 5,
    "grade_cutoffs": "Pass is 0.6",
    "course_errors": [],
    "studio_url": "//localhost:18010/course/course-v1:edX+DemoX+Demo_Course",
    "permissions": {
        "admin": true,
        "instructor": false,
        "finance_admin": false,
        "sales_admin": false,
        "staff": false,
        "forum_admin": false,
        "data_researcher": true
    },
    "tabs": [
        {
            "tab_id": "course_info",
            "title": "Course Info",
            "url": "http://localhost:8080/instructor/course-v1:edX+DemoX+Demo_Course/course_info"
        },
        {
            "tab_id": "enrollments",
            "title": "Enrollments",
            "url": "http://localhost:8080/instructor/course-v1:edX+DemoX+Demo_Course/enrollments"
        },
        {
            "tab_id": "course_team",
            "title": "Course Team",
            "url": "http://localhost:8080/instructor/course-v1:edX+DemoX+Demo_Course/course_team"
        },
        {
            "tab_id": "grading",
            "title": "Grading",
            "url": "http://localhost:8080/instructor/course-v1:edX+DemoX+Demo_Course/grading"
        },
        {
            "tab_id": "date_extensions",
            "title": "Date Extensions",
            "url": "http://localhost:8080/instructor/course-v1:edX+DemoX+Demo_Course/date_extensions"
        },
        {
            "tab_id": "data_downloads",
            "title": "Data Downloads",
            "url": "http://localhost:8080/instructor/course-v1:edX+DemoX+Demo_Course/data_downloads"
        },
        {
            "tab_id": "open_responses",
            "title": "Open Responses",
            "url": "http://localhost:8080/instructor/course-v1:edX+DemoX+Demo_Course/open_responses"
        },
        {
            "tab_id": "cohorts",
            "title": "Cohorts",
            "url": "http://localhost:8080/instructor/course-v1:edX+DemoX+Demo_Course/cohorts"
        },
        {
            "tab_id": "special_exams",
            "title": "Special Exams",
            "url": "http://localhost:8080/instructor/course-v1:edX+DemoX+Demo_Course/special_exams"
        }
    ],
    "disable_buttons": false,
    "analytics_dashboard_message": "To gain insights into student enrollment and participation <a href=\"None/courses/course-v1:edX+DemoX+Demo_Course\" rel=\"noopener\" target=\"_blank\">visit Your Platform Name Here Insights, our new course analytics product</a>."
}`

Api url:
`/api/instructor/v2/courses/course-v1:edX+DemoX+Demo_Course/instructor_tasks`
Response:
`{
    "tasks": [
        {
            "task_id": "2519ff31-22d9-4a62-91e2-55495895b355",
            "task_type": "grade_problems",
            "task_state": "PROGRESS",
            "status": "Incomplete",
            "created": "2019-01-15T18:00:15.902470+00:00",
            "task_input": "{}",
            "task_output": null,
            "duration_sec": "unknown",
            "task_message": "No status information available",
            "requester": "staff"
        }
    ]
}
`

jawad-khan committed:
feat: Instructor course information v2 apis

fix: fixed quality tests

awais786 commented on lms/djangoapps/instructor/views/api_v2.py:
```diff
@@ -0,0 +1,304 @@
+                status=status.HTTP_404_NOT_FOUND
+            )
+
+        if not request.user.has_perm(permissions.VIEW_DASHBOARD, course):
```
It can be converted into `DRF permission class`

awais786 commented on lms/djangoapps/instructor/views/api_v2.py:
```diff
@@ -0,0 +1,304 @@
+            )
+
+        # Check permissions
+        if not request.user.has_perm(permissions.SHOW_TASKS, course):
```
same here

awais786 commented on lms/djangoapps/instructor/views/api_v2.py:
```diff
@@ -0,0 +1,304 @@
+
+
+@view_auth_classes()
+class InstructorTaskListView(DeveloperErrorViewMixin, APIView):
```
Here you need to add some permission classes also e.g

`permission_classes = [CanViewInstructorDashboard]`

diana-villalvazo-wgu commented on lms/djangoapps/instructor/views/api_v2.py:
```diff
@@ -0,0 +1,304 @@
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from lms.djangoapps.courseware.tabs import get_course_tab_list
```
tabs needed for instructor dashboard are different than course tabs, afaik we need to create the logic for instructor tabs

<img width="815" height="206" alt="Image" src="https://github.com/user-attachments/assets/d61bb88c-bb25-481e-a70d-bd56e2e8c200" />

awais786 replied:
This PR is only for backend.

diana-villalvazo-wgu replied:
I know but im showing the difference on the tab list names, this PR is retrieving tab list from course tabs, and the correct list needed in the response is instructor tabs, so that logic needs to be added/created as part of this PR

jawad-khan replied:
@diana-villalvazo-wgu I’ve updated the example response with the revised tabs list — please confirm if these match what you need.

I couldn’t find URLs for all of the tabs, so I’ve included the `is_hidden` field instead, as suggested by Jesse in the original document. If you’d prefer to have a `url` field, could you please share the URLs for all the tabs so I can update them accordingly?

jawad-khan committed:
fix: fixed tabs content and refactored code

jawad-khan committed:
fix: fixed quality tests

diana-villalvazo-wgu commented on lms/djangoapps/instructor/views/serializers_v2.py:
```diff
@@ -0,0 +1,325 @@
+            },
+            {
+                'tab_id': 'discussions_management',
+                'title': _('Discussions'),
```
this one is not part of instructor tabs

jawad-khan replied:
ok, I'll remove it.

diana-villalvazo-wgu commented on lms/djangoapps/instructor/views/serializers_v2.py:
```diff
@@ -0,0 +1,325 @@
+            'data_researcher': request.user.has_perm(permissions.CAN_RESEARCH, course_key),
+        }
+
+        sections = [
```
instead of sections, can we use wording: tabs?

jawad-khan replied:
In response we are already sending tabs, I'll rename in code to tabs too in next commit.

diana-villalvazo-wgu replied:
we don't want to use is_hidden attribute, we should follow same pattern used on other places like course tabs,
Instructor tabs response should be something like this:

> tabs= [
>     {
>         "tab_id": "course_info",
>         "title": "Course Info",
>         "url": "/instructor/:courseId/course_info"
>     },
>     {
>         "tab_id": "enrollments",
>         "title": "Enrollments",
>         "url": "/instructor/:courseId/enrollments"
>     },
>     {
>         "tab_id": "grading",
>         "title": "Grading",
>         "url": "/instructor/:courseId/grading"
>     },
>     {
>         "tab_id": "course_team",
>         "title": "Course Team",
>         "url": "/instructor/:courseId/course_team"
>     },
>     {
>         "tab_id": "date_extensions",
>         "title": "Date Extensions",
>         "url": "/instructor/:courseId/date_extensions"
>     },
>     {
>         "tab_id": "data_downloads",
>         "title": "Data Downloads",
>         "url": "/instructor/:courseId/data_downloads"
>     },
>     {
>         "tab_id": "open_responses",
>         "title": "Open Responses",
>         "url": "/instructor/:courseId/open_responses"
>     },
>     {
>         "tab_id": "certificates",
>         "title": "Certificates",
>         "url": "/instructor/:courseId/certificates"
>     },
>     {
>         "tab_id": "cohorts",
>         "title": "Cohorts",
>         "url": "/instructor/:courseId/cohorts"
>     },
>     {
>         "tab_id": "special_exams",
>         "title": "Special Exams",
>         "url": "/instructor/:courseId/special_exams"
>     }
> ]


Note: I added the url paths, but the complete url needs to be built in the backend, for example for course info if you are on dev environment and course Id is `course-v1:OpenedX DemoX DemoCourse` , response should return:

```
    {
        "tab_id": "course_info",
        "title": "Course Info",
        "url": "http://local.openedx.io:8000/instructor/course-v1:OpenedX+DemoX+DemoCourse/course_info"
    }
```

jawad-khan replied:
@diana-villalvazo-wgu 
How will the frontend determine which tabs to display or hide if the is_hidden flag is not included in the API response?

diana-villalvazo-wgu replied:
tabs response should be dynamic, the backend should decide which tabs to send to be showed on the frontend, thats what i meant that the logic for the tabs still need to be built as part of this PR

jawad-khan replied:
@diana-villalvazo-wgu
Could you please share valid URLs for all the tabs so I can populate them in the url field? The ones provided earlier don’t appear to be valid.

Also, I noticed that some tabs — for example, Membership — are missing from your example response. Could you please include url for membership as well?

diana-villalvazo-wgu replied:
those are all the tabs with pathnames, why those urls are not valid?
some of them were renamed as part of this effort like membership was changed to course team and student admin was changed to grading.

wgu-jesse-stewart replied:
@jawad-khan 
   
tabs:
```
[
        {
            "tab_id": "course_info",
            "title": "Course Info",
            "url": f"{base_url}/instructor/{course_id}/course_info"
        },
        {
            "tab_id": "enrollments",
            "title": "Enrollments",
            "url": f"{base_url}/instructor/{course_id}/enrollments"
        },
        {
            "tab_id": "grading",
            "title": "Grading",
            "url": f"{base_url}/instructor/{course_id}/grading"
        },
        {
            "tab_id": "course_team",
            "title": "Course Team",
            "url": f"{base_url}/instructor/{course_id}/course_team"
        },
        {
            "tab_id": "date_extensions",
            "title": "Date Extensions",
            "url": f"{base_url}/instructor/{course_id}/date_extensions"
        },
        {
            "tab_id": "data_downloads",
            "title": "Data Downloads",
            "url": f"{base_url}/instructor/{course_id}/data_downloads"
        },
        {
            "tab_id": "open_responses",
            "title": "Open Responses",
            "url": f"{base_url}/instructor/{course_id}/open_responses"
        },
        {
            "tab_id": "certificates",
            "title": "Certificates",
            "url": f"{base_url}/instructor/{course_id}/certificates"
        },
        {
            "tab_id": "cohorts",
            "title": "Cohorts",
            "url": f"{base_url}/instructor/{course_id}/cohorts"
        },
        {
            "tab_id": "special_exams",
            "title": "Special Exams",
            "url": f"{base_url}/instructor/{course_id}/special_exams"
        }
    ]
```
    
Tab Permissions:


Section | Code Reference | Conditions to Display | Notes / Description
-- | -- | -- | --
Course Info | instructor_dashboard.py:451 | access['staff'] |  
Enrollments (previously: Membership) | instructor_dashboard.py:499 | access['staff'] | Manages student enrollment.
Grading (previously: Student Admin) | instructor_dashboard.py:571 | access['staff'] | Student-specific administrative functions.
Course Team (previously: Membership) | instructor_dashboard.py:499 | access['staff'] | Manages beta testers, course roles, and forum members.
Date Extensions | instructor_dashboard.py:620 | access['instructor'] AND is_enabled_for_course(course_key) (edx-when feature) | Manages due date extensions.
Data Downloads | instructor_dashboard.py:638 | access['data_researcher'] | Hidden if user lacks access. Section key varies: data_download_2 or data_download depending on v2 toggle.
Open Responses | instructor_dashboard.py:769 | Course has non-orphaned ORA blocks AND access['staff'] | Open Response Assessment (ORA) management.
Certificates | instructor_dashboard.py:341 | CertificateGenerationConfiguration.current().enabled AND not CCX AND (access['admin'] OR (access['instructor'] AND feature flag enabled)) | Certificate generation and management.
Cohorts | instructor_dashboard.py:534 | access['staff'] | Cohort management.
Special Exams | instructor_dashboard.py:307 | (proctored OR timed exams enabled) AND user is Staff/CourseStaffRole/CourseInstructorRole AND feature flag enabled | Proctored and timed exam management.
Discussions (not in new UI) | instructor_dashboard.py:553 | access['staff'] AND legacy discussions enabled. Hidden if not cohorted AND no enrollment track divisions. | Discussion topic and settings management.
Analytics (not in new UI) | instructor_dashboard.py:758 | Analytics dashboard configured AND (access['staff'] OR access['instructor']) | Links to external analytics dashboard.
Email (not in new UI) | instructor_dashboard.py:696 | Bulk email enabled feature flag AND not disabled for course AND (access['staff'] OR access['instructor']) | Bulk email functionality.

jawad-khan replied:
@wgu-jesse-stewart My logic is based on the same code reference you mentioned.
I just have one clarification — is the base_url supposed to point to the MFE or the LMS instance?

jawad-khan replied:
@wgu-jesse-stewart @diana-villalvazo-wgu 
I have added my changes based on lms instance, let me know if you it other way or just relative path. I have also updated example response in PR description.

diana-villalvazo-wgu replied:
base_url should be the MFE, since this are the actual urls for pages in MFE, not lms endpoints

jawad-khan replied:
I updated base url and its now pointing to mfe. Updated response too. 
You have to set `INSTRUCTOR_MICROFRONTEND_URL` in lms settings and it wiil work for you.

jawad-khan committed:
fix: dynamic add tabs

fix: refactored tabs logic

jawad-khan committed:
fix: fixed mfe base url

diana-villalvazo-wgu commented on lms/djangoapps/instructor/views/api_urls.py:
```diff
@@ -20,6 +20,19 @@
+    re_path(
+        rf'^courses/{COURSE_ID_PATTERN}$',
+        api_v2.CourseMetadataView.as_view(),
+        name='course_metadata'
```
do we need this? 👀

jawad-khan replied:
yes.

feanil replied:
@diana-villalvazo-wgu are you asking about the name specifically? If that's the case, we name the endpoints because it's handy to be able to use the [reverse](https://docs.djangoproject.com/en/5.2/ref/urlresolvers/#reverse) feature of django to fetch URLs.

jawad-khan commented:
@feanil This PR appears to be final from the requirements side. Can you please approve this.
It introduces a new API implemented as [recommended by you](https://github.com/openedx/edx-platform/issues/37465#issuecomment-3469308327).
Since it’s a new addition, it won’t impact any existing functionality and will take effect once we start integrating it into the MFE.

wgu-jesse-stewart reviewed (changes requested):
Hey [@jawad-khan](https://github.com/jawad-khan), thanks for tackling this! The v2 API implementation is looking really solid. I like how you've structured the serializers and the test coverage is comprehensive. Nice work refactoring that analytics dashboard message into a reusable helper too.
Just a few things to clean up:

Can we match the Figma design order here? Should be:

- Course Info
- Enrollments
- Course Team
- Grading
- Date Extensions
- Data Downloads
- Open Responses
- Certificates
- Cohorts
- Special Exams

According to Figma, this should be data_downloads (plural with underscore) and the title should be Data Downloads (with space). Looks like the current response has data_download (singular).

The tabs URLs in the example response are missing the INSTRUCTOR_MICROFRONTEND_URL prefix. Might want to update those to show the full URL format like you mentioned needs to be set in LMS settings. Can you also use real examples that the endpoint will return? EG: Course Info, Enrollments.

This is a great foundation for the instructor MFE integration. Once these tweaks are in, we should be good to go!

jawad-khan committed:
fix: fixed tabs order

jawad-khan committed:
fix: fixed test cases

jawad-khan commented:
> Hey [@jawad-khan](https://github.com/jawad-khan), thanks for tackling this! The v2 API implementation is looking really solid. I like how you've structured the serializers and the test coverage is comprehensive. Nice work refactoring that analytics dashboard message into a reusable helper too. Just a few things to clean up:
> 
> Can we match the Figma design order here? Should be:
> 
>     * Course Info
> 
>     * Enrollments
> 
>     * Course Team
> 
>     * Grading
> 
>     * Date Extensions
> 
>     * Data Downloads
> 
>     * Open Responses
> 
>     * Certificates
> 
>     * Cohorts
> 
>     * Special Exams
> 
> 
> According to Figma, this should be data_downloads (plural with underscore) and the title should be Data Downloads (with space). Looks like the current response has data_download (singular).
> 
> The tabs URLs in the example response are missing the INSTRUCTOR_MICROFRONTEND_URL prefix. Might want to update those to show the full URL format like you mentioned needs to be set in LMS settings. Can you also use real examples that the endpoint will return? EG: Course Info, Enrollments.
> 
> This is a great foundation for the instructor MFE integration. Once these tweaks are in, we should be good to go!

- I have ordered the list as much as code allowed me but you shouldn't rely on api list ordering.
- Fixed data downloads issue
- tabs url has url prefix, look to the sample response given in PR description.
- This response is from a Demo course live on my local environment.

jawad-khan committed:
fix: fixed quality tests

wgu-jesse-stewart commented on lms/djangoapps/instructor/views/api_v2.py:
```diff
@@ -0,0 +1,271 @@
+                {
+                  "tab_id": "courseware",
+                  "title": "Course",
+                  "url": "/courses/course-v1:edX+DemoX+2024/courseware"
```
@jawad-khan This comment doesn't accurately reflect what is being returned by the API

wgu-jesse-stewart commented:
@jawad-khan Thanks for the updates! Quick follow-up on the tab ordering - we actually do need to rely on the API list ordering since the MFE will render tabs in the order they're received.

What do you think about this approach to fix the sort order? Instead of adding a sort_order field, we can just ensure the tabs are appended in the correct sequence. I think returning them in the right order from the API is cleaner than making the frontend handle sorting.

Here's what I'm thinking:
<img width="774" height="513" alt="image" src="https://github.com/user-attachments/assets/4af2fb04-1387-4e7c-992b-5b7bcb5dc85f" />
<img width="948" height="344" alt="image" src="https://github.com/user-attachments/assets/372bf161-2663-4e05-bd37-785d4e819c83" />

This way the tabs will always come back in the Figma-specified order without any extra complexity on the frontend side.
Thoughts?

feanil commented on lms/djangoapps/instructor/views/api_v2.py:
```diff
@@ -0,0 +1,271 @@
+        course_key = CourseKey.from_string(course_id)
+        course = get_course_by_id(course_key)
+
+        tabs = get_course_tab_list(request.user, course)
```
What @diana-villalvazo-wgu is saying above is that this is the wrong list of tabs to return.  These are the course tabs, and not the Instructor API tabs.  That list is compiled in the [instructor_dashboard_2](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/instructor/views/instructor_dashboard.py#L135) function.  You'll want to maybe need to extract that logic for the sections context into a shared function that both that function and this API endpoint call.

jawad-khan replied:
I have added from [instructor_dashboard_2](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/instructor/views/instructor_dashboard.py#L135) and fixed their order according to the comment. Please give this PR a second pass.

tonybusa commented:
Hi @jawad-khan - just checking in here, this PR is blocking critical and timely work by our team. Are you planning to work on it this week, or would you mind if our team took it over?

cc: @diana-villalvazo-wgu , @wgu-jesse-stewart , @sarina

jawad-khan committed:
fix: fixed tab order

jawad-khan commented:
@wgu-jesse-stewart I have fixed tabs order and updated url in comment. Let mw know if anything else is needed. 

@tonybusa PR is final from my side, we just need a thumbs up to merge it. All requirements mentioned by @wgu-jesse-stewart and @diana-villalvazo-wgu have been implemented.

wgu-jesse-stewart reviewed (approved):
Thanks for taking the time to polish this endpoint, @jawad-khan! Everything is working as expected now. Great work getting this across the finish line

feanil commented on lms/djangoapps/instructor/views/serializers_v2.py:
```diff
@@ -0,0 +1,337 @@
+                'url': f'{mfe_base_url}/instructor/{str(course_key)}/special_exams'
+            })
+
+        # Frontend requires tabs to be in a specific order
```
```suggestion
        # We provide the tabs in a specific order based on how it was
        # historically presented in the frontend.  The frontend can use
        # this info or choose to ignore the ordering.
```

Remember that this API might outlive this implementation of the frontend and so more context here is useful to provide.

At first I thought that this ordering info should actually live in the frontend but then I realized that it would cause some coupling where if we ever change the names of these tabs in the backend, we'd have to synchronize the update to the frontend which doesn't make sense.  So I'm fine with leaving this here but wanted to suggest a better comment.

jawad-khan replied:
I have updated comment and rebased this PR from master.

feanil reviewed (approved):
Generally looks good to me, one small suggestion.

jawad-khan committed:
fix: updated comment
